dali/operators/generic/lookup_table.cu

```diff
@@ -24,57 +25,67 @@ namespace detail {
 constexpr auto kMaxKey = LookupTable<GPUBackend>::kMaxKey;
 
 template <typename OutputType, typename InputType>
-__global__ void
-LookupValuesImpl(OutputType *output,
-                 const InputType *input,
-                 size_t data_size,
-                 const OutputType* lookup_table,
-                 const OutputType default_value) {
+__global__ void LookupValuesImpl(const LutSampleDesc *samples, const kernels::BlockDesc<1> *blocks,
+                                 const OutputType *lookup_table, const OutputType default_value) {
   // We do not check the key range when the type range is smaller than the supported range
   constexpr bool check_range =
-       !std::is_same<InputType, uint8_t>::value
-    && !std::is_same<InputType, uint16_t>::value;
+      !std::is_same<InputType, uint8_t>::value && !std::is_same<InputType, uint16_t>::value;
   constexpr auto max_key = ConvertSat<InputType>(kMaxKey);
 
-  size_t tid = blockDim.x * blockIdx.x + threadIdx.x;
-  if (tid >= data_size)
-    return;
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
 
-  const auto key = input[tid];
-  if (check_range) {
-    output[tid] = (std::is_unsigned<InputType>::value || key >= 0) && key <= max_key ?
-      lookup_table[key] : default_value;
-  } else {
-    output[tid] = lookup_table[key];
+  auto *output = reinterpret_cast<OutputType *>(sample.output);
+  const auto *input = reinterpret_cast<const InputType *>(sample.input);
+  for (int x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
+    const auto key = input[x];
+    if (check_range) {
+      output[x] = (std::is_unsigned<InputType>::value || key >= 0) && key <= max_key ?
+                      lookup_table[key] :
+                      default_value;
+    } else {
+      output[x] = lookup_table[key];
+    }
   }
 }
 
 }  // namespace detail
 
 template<>
 void LookupTable<GPUBackend>::RunImpl(DeviceWorkspace &ws) {
-  const auto &input = ws.Input<GPUBackend>(0);
-  auto &output = ws.Output<GPUBackend>(0);
+  const auto &input = ws.InputRef<GPUBackend>(0);
+  const auto &shape = input.shape();
+  auto &output = ws.OutputRef<GPUBackend>(0);
   output.SetLayout(input.GetLayout());
-  auto data_size = input.size();
-  constexpr int kThreads = 512;
-  const int blocks = (data_size + kThreads - 1) / kThreads;
+
   const auto stream = ws.stream();
-  Tensor<GPUBackend> lookup_table_gpu;
+
+  auto num_samples = shape.num_samples();
+  samples_.resize(num_samples);
+  for (int sample_id = 0; sample_id < num_samples; sample_id++) {
+    samples_[sample_id].output = output.raw_mutable_tensor(sample_id);
+    samples_[sample_id].input = input.raw_tensor(sample_id);
+  }
+  samples_dev_.from_host(samples_, stream);
+
+  std::array<std::pair<int, int>, 1> collapse_groups = {{{0, shape.sample_dim()}}};
+  auto collapsed_shape = collapse_dims<1>(shape, collapse_groups);
```

Comment:
Nitpick - I guess you could get away with:
```suggestion
  auto collapsed_shape = collapse_dims<1>(shape, { make_pair(0, shape.sample_dim()) });
```

Reply:
done